Activities/Database/UiPath.Database/DatabaseConnection.cs

```diff
@@ -94,6 +99,182 @@ public virtual int InsertDataTable(string tableName, DataTable dataTable)
             return dbDA.Update(dataTable);
         }
 
+        
+        public long BulkInsertDataTable(string tableName, DataTable dataTable, string connection)
+        {
+            DbDataAdapter dbDA = DbProviderFactories.GetFactory(_providerName).CreateDataAdapter();
+            DbCommandBuilder cmdb = DbProviderFactories.GetFactory(_providerName).CreateCommandBuilder();
+            cmdb.DataAdapter = dbDA;
+            dbDA.ContinueUpdateOnError = false;
+
+            long affectedRecords = 0;
+
+           
+
+            if (_providerName == "System.Data.SqlClient")
+            {
+                //if SQL Server
+
+                dbDA.SelectCommand = _connection.CreateCommand();
+                dbDA.SelectCommand.Transaction = _transaction;
+                dbDA.SelectCommand.CommandType = CommandType.Text;
+                dbDA.SelectCommand.CommandText = string.Format("select TOP(1) * from {0}", tableName);
+
+                var ds = new DataSet();
+                dbDA.Fill(ds);
+                ValidateTableStructure(tableName, dataTable, ds);
+
+                // Perform an initial count on the destination table.
+                SqlCommand commandRowCount = new SqlCommand(string.Format("SELECT COUNT(*) FROM {0}", tableName), (SqlConnection)_connection);
+                long countStart = Convert.ToInt32(
+                    commandRowCount.ExecuteScalar());
+                // Set up the bulk copy object
+                using (SqlBulkCopy bulkCopy = new SqlBulkCopy(_connection.ConnectionString, SqlBulkCopyOptions.UseInternalTransaction))
```

Comment:
not sure if we can moq the SqlBulkCopy object directly but if we wrap this in a factory method and return an interface we could write some tests for the rest of the code

Reply:
extracted to interface in a separate file.